src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -1700,6 +1701,182 @@ impl ToQemuParams for ObjectIoThread {
     }
 }
 
+// PCIeRootPortDevice directly attached onto the root bus
+// -device pcie-root-port,id=rp0,bus=pcie.0,chassis=0,slot=0,multifunction=off,pref64-reserve=<X>B,mem-reserve=<Y>B
+#[derive(Debug, Default)]
+pub struct PCIeRootPortDevice {
+    id: String,
+    bus: String,
+    chassis: String,
+    slot: String,
+    multifunction: bool,
+    addr: String,
+
+    mem_reserve: String,
+    pref64_reserve: String,
+}
+
+impl PCIeRootPortDevice {
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        id: &str,
+        bus: &str,
+        chassis: &str,
+        slot: &str,
+        multifunc: bool,
+        addr: &str,
+        mem_reserve: u64,
+        pref64_reserve: u64,
+    ) -> Self {
+        PCIeRootPortDevice {
+            id: id.to_string(),
+            bus: bus.to_string(),
+            chassis: chassis.to_string(),
+            slot: slot.to_string(),
+            multifunction: multifunc,
+            addr: addr.to_string(),
+            mem_reserve: format!("{}B", mem_reserve),
+            pref64_reserve: format!("{}B", pref64_reserve),
+        }
+    }
+}
+
+#[async_trait]
+impl ToQemuParams for PCIeRootPortDevice {
+    async fn qemu_params(&self) -> Result<Vec<String>> {
+        let mut qemu_params = Vec::new();
+        let mut device_params = Vec::new();
```

Comment:
Is there any advantage to using two vectors over the idiom used elsewhere in the file which allocates a single vector at the top of the function
```
let mut params = Vec::new();
```
and finishes after filling it in with
```
Ok(vec!["-device".to_owned(), params.join(",")])
```
?